[005] - Add CDC Join tutorial and fix other minor issues (#2387)
hazelcast/hazelcast-jet: site/docs/tutorials/cdc-join.md

Here we will examine how to make our map hold enriched data, combined
(joined, if you will) from multiple tables. Let's, for example, make
each of our enriched data objects hold information about one customer
and all the orders belonging to them.

## 1. Install Docker

This tutorial uses [Docker](https://www.docker.com/) to simplify the
setup of databases, which you can freely experiment on.

1. Follow Docker's [Get Started](https://www.docker.com/get-started)
   instructions and install it on your system.
2. Test that it works:
   * Run `docker version` to check that you have the latest release
     installed.
   * Run `docker run hello-world` to verify that Docker is pulling
     images and running as expected.

## 2. Start Database

Exact instructions for starting the supported databases are as follows:

* [MySQL](cdc.md#2-start-mysql-database)
* [Postgres](cdc-postgres.md#2-start-postgresql-database).

## 3. Start Command Line Client

Exact instructions for starting the specific command line client of the
supported databases are as follows:

* [MySQL](cdc.md#3-start-mysql-command-line-client)
* [Postgres](cdc-postgres.md#3-start-postgresql-command-line-client)

> ## 4. Create a New Java Project

Reviewer: The linked tutorial says to create `cdc-tutorial` project. 

With maven it is fine, there is the artifactId specified.
With gradle it is not defined in the build.gradle file but in settings.gradle and the value defaults to the directory name.

This should be highlighted, because you refer to the jar later. (I created folder called cdc-join and had to change the user code deployment and submit command).

---
title: Join Change Data Capture Records
sidebar_label: Join
description: How to maintain a joined image of Change Data Capture data from multiple tables.
---

**Change data capture** refers to the process of **observing changes
made to a database** and extracting them in a form usable by other
systems, for the purposes of replication, analysis and many more.

Change Data Capture is especially important to Jet, because it allows

> for the **streaming of changes from databases**, which can be

Reviewer: ```suggestion
**streaming changes from a database**, which can be
```

Author: In my opinion the correction is not grammatically correct and does not improve the original version.

efficiently processed by Jet.

Implementation of CDC in Jet is based on
[Debezium](https://debezium.io/), which is an open source distributed
platform for change data capture.

In our previous tutorials on how to extract change event data from

> [MySQL](cdc-mysql.md) and [Postgres](cdc-postgres.md) databases we've

Reviewer: There should be a comma somewhere :-).
```suggestion
[MySQL](cdc-mysql.md) and [Postgres](cdc-postgres.md) databases, we've
```

> [MySQL](cdc-mysql.md) and [Postgres](cdc-postgres.md) databases we've

Reviewer 2: shoudl be `cdc`

> seen how to monitor single database tables and how maintain an

Reviewer: ```suggestion
seen how to monitor a single database table and how to maintain an
```

> up-to-date image of them in memory, as a map.

Reviewer: ```suggestion
up-to-date image of the table in memory, in a form of a map.
```

Here we will examine how to make our map hold enriched data, combined
(joined, if you will) from multiple tables. Let's say customers and all
the orders belonging to them, as one entity.

## 1. Install Docker

This tutorial uses [Docker](https://www.docker.com/) to simplify the

> setup of a PostgreSQL database, which you can freely experiment on.

Reviewer: Is this for just postgres or both postgres and mysql?

Author: Good catch, copy-paste error.

1. Follow Docker's [Get Started](https://www.docker.com/get-started)
   instructions and install it on your system.
2. Test that it works:
   * Run `docker version` to check that you have the latest release
     installed.
   * Run `docker run hello-world` to verify that Docker is pulling
     images and running as expected.

## 2. Start Database

> In this tutorial we will provide all code and configs for MySQL, but

Reviewer: Seems to contain both?

Author: Yeah, I was thorn between covering both or focusing on one of them and this is a left-over. Will fix.

adapting them to any of the databases covered by the previous tutorials
should be pretty straightforward.

Exact instructions for starting the supported databases are as follows:

> * [MySQL](cdc-mysql.md#2-start-mysql-database)

Reviewer 2: should be `cdc`

    private final Map<Integer, Order> orders;
    private Customer customer;

    public OrdersOfCustomer() {
        this.customer = null;
        this.orders = new HashMap<>();
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }

    public void deleteOrder(Order order) {
        orders.remove(order.getOrderNumber());
    }

    public void addOrUpdateOrder(Order order) {
        orders.put(order.getOrderNumber(), order);
    }

    @Override
    public int hashCode() {
        return Objects.hash(customer, orders);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        OrdersOfCustomer other = (OrdersOfCustomer) obj;
        return Objects.equals(customer, other.customer)
                && Objects.equals(orders, other.orders);
    }

    @Override
    public String toString() {
        return String.format("Customer: %s, Orders: %s", customer, orders);
    }
}
```

Careful observers will notice another thing going on, an additional

> processing stage which handles and fixes event reoreding that might

Reviewer: ```suggestion
processing stage which handles and fixes event reordering that might
```

```

This will produce a jar file called `cdc-tutorial-1.0-SNAPSHOT.jar`
in the `build/libs` folder of our project.

<!--Maven-->

```bash
mvn package
```

This will produce a jar file called `cdc-tutorial-1.0-SNAPSHOT.jar`
in the `target` folder or our project.

<!--END_DOCUSAURUS_CODE_TABS-->

## 7. Start Hazelcast Jet

1. Download Hazelcast Jet

```bash
wget https://github.com/hazelcast/hazelcast-jet/releases/download/v{jet-version}/hazelcast-jet-{jet-version}.tar.gz
tar zxvf hazelcast-jet-{jet-version}.tar.gz && cd hazelcast-jet-{jet-version}
```

2. Activate the CDC plugin for the database you're using:

<!--DOCUSAURUS_CODE_TABS-->

<!--MySQL-->

```bash
mv opt/hazelcast-jet-cdc-debezium-{jet-version}.jar lib; \
mv opt/hazelcast-jet-cdc-mysql-{jet-version}.jar lib
```

<!--Postgres-->

```bash
mv opt/hazelcast-jet-cdc-debezium-{jet-version}.jar lib; \
mv opt/hazelcast-jet-cdc-postgres-{jet-version}.jar lib
```

<!--END_DOCUSAURUS_CODE_TABS-->

3. Enable user code deployment:

Due to the type of sink we have used in our pipeline we need to make
some extra changes in order for the IMDG cluster (which Jet sits
on top of) be aware of the custom classes we have defined.

> Pls. append following config lines to the `config/hazelcast.yaml` file,

Reviewer: ```suggestion
Please append the following config lines to the `config/hazelcast.yaml` file,
```
